joingroup.go

```diff
@@ -0,0 +1,212 @@
+package kafka
+
+import (
+	"bufio"
+	"bytes"
+)
+
+type memberProtocolMetadata struct {
+	// MemberID assigned by the group coordinator or null if joining for the
+	// first time.
+	MemberID string
+	Metadata protocolMetadata
+}
+
+type protocolMetadata struct {
+	Version  int16
+	Topics   []string
+	UserData []byte
+}
+
+func (t protocolMetadata) size() int32 {
+	return sizeofInt16(t.Version) +
+		sizeofStringArray(t.Topics) +
+		sizeofBytes(t.UserData)
+}
+
+func (t protocolMetadata) writeTo(w *bufio.Writer) {
+	writeInt16(w, t.Version)
+	writeStringArray(w, t.Topics)
+	writeBytes(w, t.UserData)
+}
+
+func (t protocolMetadata) bytes() []byte {
```

Comment:
Those `bytes() []byte` methods don't seem to be used anywhere, which means they are also untested. Can we get rid of them and only introduce them if we need to?

Reply:
```bytes()``` is used by newJoinGroupRequestV2 and is needed as part of the consumer group handshake to declare the participation in the consumer group.

Reply:
Ah! I searched the page but Github hid some larger files so I missed it. All good 👍

Reply:
Actually I can't find `newJoinGroupRequestV2`, the only place I see it used is in a test https://github.com/segmentio/kafka-go/pull/47/files#diff-9d43d24e28e45a8fd8a0ffe03a0e925aR17

Am I missing something?

Reply:
Ah right, I'm looking on different branch. I'll go ahead and backport the tests.